src/filter/bilateral.rs

```diff
@@ -0,0 +1,303 @@
+        } else {
+            pixel1
+                .channels()
+                .iter()
+                .zip(pixel2.channels().iter())
+                .map(|(c1, c2)| (f32::from(*c1) - f32::from(*c2)).powi(2))
+                .sum::<f32>()
+                .sqrt()
+        };
+
+        gaussian_weight(euclidean_distance, self.sigma)
+    }
+}
+
+/// Denoise an 8-bit image while preserving edges using bilateral filtering.
+///
+/// # Arguments
+///
+/// * `image` - Image to be filtered.
+/// * `window_size` - Window size for filtering.
+/// * `spatial_sigma` - Standard deviation for euclidean spatial distance. A larger value results in
+///     averaging of pixels with larger spatial distances.
+/// * `color_similarity` - An type which implements [`ColorSimilarity`] returns the color
+///     similarity between two pixels. Common examples may include simple absolute difference for
+///     greyscale pixels or cartesian distance in the CIE-Lab color space \[1\].
+///
+/// This filter averages pixels based on their spatial similarity as well as their color
+/// similarity. Spatial similarity is measured by the Gaussian function of the Euclidean distance
+/// between two pixels with the user-specified standard deviation (`sigma_spatial`). Color
+/// similarity is provided via the `color_similarity` argument.
+///
+/// # References
+///
+///   \[1\] C. Tomasi and R. Manduchi. "Bilateral Filtering for Gray and Color
+///        Images." IEEE International Conference on Computer Vision (1998)
+///        839-846. DOI: 10.1109/ICCV.1998.710815
+///
+/// # Panics
+///
+/// 1. If `image.width() > i32::MAX as u32`
+/// 2. If `image.height() > i32::MAX as u32`.
+/// 3. If `image.width() == 0`
+/// 4. If `image.height() == 0`
+///
+/// # Examples
+///
+/// ```
+/// use imageproc::filter::bilateral::{bilateral_filter, GaussianEuclideanColorSimilarity};
+/// use imageproc::utils::gray_bench_image;
+///
+/// let image = gray_bench_image(50, 50);
+///
+/// let filtered = bilateral_filter(&image, 2, 3., GaussianEuclideanColorSimilarity { sigma: 10.0 });
+/// ```
+#[must_use = "the function does not modify the original image"]
+pub fn bilateral_filter<I, P, C>(
+    image: &I,
+    radius: u8,
+    spatial_sigma: f32,
+    color_similarity: C,
+) -> Image<P>
+where
+    I: GenericImage<Pixel = P>,
+    P: WeightedAverage + image::Pixel,
+    C: ColorSimilarity<P>,
+{
+    assert!(!image.width() > i32::MAX as u32);
+    assert!(!image.height() > i32::MAX as u32);
+    assert_ne!(image.width(), 0);
+    assert_ne!(image.height(), 0);
+
+    let radius = i16::from(radius);
+
+    let window_range = -radius..=radius;
+    let spatial_similarity_lookup = window_range
+        .clone()
+        .cartesian_product(window_range.clone())
+        .map(|(w_y, w_x)| {
+            //The gaussian of the euclidean spatial distance
+            gaussian_weight(
+                (f32::from(w_x).powi(2) + f32::from(w_y).powi(2)).sqrt(),
```

Comment:
Same comment as for `color_similarity` above.

Reply:
I've switched back to distance here too